DOC-8059 changed n1ql subqueries to support 7.0 version (#1905)
couchbase/docs-server: modules/n1ql/pages/n1ql-language-reference/subquery-examples.adoc

> ----

Reviewer: First, please add the following admonition _before_ the first Example heading:

``` asciidoc
WARNING: Please note that the examples below will alter the data in your sample buckets.
To restore your sample data, remove and reinstall the `travel-sample` bucket.
Refer to xref:manage:manage-settings/install-sample-buckets.adoc[Sample Buckets] for details.
```

Secondly, please change the language for the first code block to N1QL:

``` asciidoc
[source,n1ql]
```

Reviewer: This warning applies to all the examples, not just example 1, so please move the warning to line 3 — _before_ the first Example heading

SELECT name, (SELECT raw avg(s.ratings.Overall)
              FROM   t.reviews  as s)[0] AS overall_avg_rating
FROM   `travel-sample`.inventory.hotel AS t
ORDER BY overall_avg_rating DESC
LIMIT 3;

> 

Reviewer: Separate the query and the result:

```suggestion
----

.Results
[source,json]
----
```

----
SELECT  id, sourceairport, destinationairport,
            (SELECT s.*

>              FROM `travel-sample`.inventory.route s

Reviewer: The source of the subquery should be `route` to correlate with the outer query:

```suggestion
             FROM route.schedule s
```

             WHERE s.utc > "22:00:00"
             ORDER BY s.utc)  after_10pm
FROM `travel-sample`.inventory.route
WHERE sourceairport = "SFO"
LIMIT 2;
----

> .Results:

Reviewer: ```suggestion

.Results
```

Delete trailing colon `:`

[source,json]
----
[
  {

>     "after_10pm": [],

Reviewer: When the subquery is properly correlated with the outer query, it returns the correct results:

```suggestion
    "after_10pm": [
      {
        "day": 2,
        "flight": "AI712",
        "utc": "23:44:00"
      }
    ],
```

    "destinationairport": "HKG",
    "id": 10624,
    "sourceairport": "SFO"
  },
  {

>     "after_10pm": [],

Reviewer: Again, the results of the subquery should be as follows:

```suggestion
    "after_10pm": [
      {
        "day": 6,
        "flight": "AI588",
        "utc": "22:14:00"
      },
      {
        "day": 5,
        "flight": "AI886",
        "utc": "22:53:00"
      },
      {
        "day": 1,
        "flight": "AI678",
        "utc": "23:50:00"
      }
    ],
```

    "destinationairport": "ICN",
    "id": 10625,
    "sourceairport": "SFO"
  }

> ],

Reviewer: Delete trailing comma

```suggestion
]
```

> [source,json]

Reviewer: Query is N1QL

```suggestion
[source,n1ql]
```

----
SELECT airline, sourceairport
FROM  `travel-sample`.inventory.route
WHERE (SELECT raw count(*)

>             FROM `travel-sample`.inventory.route as s WHERE s.day = 1)[0]  > 4

Reviewer: The source of the suqbuery should be `route` to correlate with the outer query

```suggestion
       FROM route.schedule as s WHERE s.day = 1)[0] > 4
```

LIMIT 3;

> 

Reviewer: Separate query and results:

```suggestion
----

.Results
[source,json]
----
```

Also update the results! The correct results should be as follows:

``` json
[
  {
    "airline": "AF",
    "sourceairport": "TLV"
  },
  {
    "airline": "AF",
    "sourceairport": "TPE"
  },
  {
    "airline": "AF",
    "sourceairport": "TRI"
  }
]
```

> [source,json]

Reviewer: Query is N1QL

```suggestion
[source,n1ql]
```

----
SELECT name, cnt_reviewers
FROM   `travel-sample`.inventory.hotel AS t
LET cnt_reviewers = (SELECT raw count(*)
                     FROM t.reviews AS s
                     WHERE s.ratings.Overall >= 4)[0]
WHERE  cnt_reviewers >= 6
ORDER BY cnt_reviewers DESC
LIMIT 3;

> 

Reviewer: Separate query and results

```suggestion
----

.Results
[source,json]
----
```

> [source,json]

Reviewer: Query is N1QL

```suggestion
[source,n1ql]
```

----
MERGE INTO `travel-sample`.inventory.hotel t USING [{"id":"21728"},{"id":"21730"}] source
ON KEY "hotel_"|| source.id
WHEN MATCHED THEN UPDATE SET t.old_vacancy = t.vacancy, t.vacancy = false

> RETURNING meta(t).id, t.old_vacancy, t.vacancy;

Reviewer: Separate query and result

```suggestion
RETURNING meta(t).id, t.old_vacancy, t.vacancy;
----

.Results
[source,json]
----
```

> [source,json]

Reviewer: Query is N1QL

```suggestion
[source,n1ql]
```

----
SELECT count(*) FROM `travel-sample`.inventory.airport t

> LET x = t

Reviewer: You must include the `geo` field otherwise the query can't find the `alt` sub-field

```suggestion
LET x = t.geo
```

WHERE (SELECT RAW y.alt FROM x y)[0] > 6000;
----

> 

Reviewer: Include results

```suggestion
.Results
[source,json]
----
[
  {
    "$1": 38
  }
]
----
```

> [source,json]

Reviewer: Add explanation of this example. Query is N1QL

```suggestion
In this example, the subquery is not correlated with the parent query, so it returns all of the airports in the `airport` collection.

[source,n1ql]
```

----
SELECT array_length((SELECT RAW t1.geo.alt
                     FROM `travel-sample`.inventory.airport t1))
FROM `travel-sample`.inventory.airport LIMIT 4;

> [

Reviewer: Separate query and results

```suggestion
FROM `travel-sample`.inventory.airport LIMIT 4;
----

.Results
[source,json]
----
[
```

Here is an example of LET variable in the FROM clause.

[source,n1ql]
----
SELECT count(*) FROM `travel-sample`.inventory.airport t
LET x = t.geo
WHERE (SELECT RAW y.alt FROM x y)[0] > 6000;
----
.Results
[source,json]
----
[
  {
    "$1": 38
  }
]
----
== Example 7

> An example of using same keyspace name in subquery FROM clause that is used in the parent query.

Reviewer: Missed the explanation for this example:

```suggestion
An example of using same keyspace name in subquery FROM clause that is used in the parent query.

In this example, the subquery is not correlated with the parent query, so it returns all of the airports in the `airport` collection.
```

> [source,json]

Reviewer: Add explanation of this example. Query is N1QL.

```suggestion
In this example, the subquery is correlated with the parent query, so it only returns the single airport found by the parent query.

[source,n1ql]
```

----
SELECT array_length((SELECT RAW t1.geo.alt FROM t t1))
FROM `travel-sample`.inventory.airport t;

> 

Reviewer: Separate query and results:

```suggestion
----

.Results
[source,json]
----
```

> [source,json]

Reviewer: Query is N1QL

```suggestion
[source,n1ql]
```

----
Update `travel-sample`.inventory.airport t1 set airportname_dup = "high_altitude_" || airportname
WHERE  t1.geo.alt IN (SELECT raw t2.geo.alt
                                           FROM `travel-sample`.inventory.airport t2
                                           WHERE t2.geo.alt > 6000)

> RETURNING *;

Reviewer: Consistent uppercase for N1QL commands, and let's return shorter results

```suggestion
UPDATE `travel-sample`.inventory.airport t1 SET airportname_dup = "high_altitude_" || airportname
WHERE  t1.geo.alt IN (SELECT RAW t2.geo.alt
                      FROM `travel-sample`.inventory.airport t2
                      WHERE t2.geo.alt > 6000)
RETURNING t1.airportname_dup;
```

----

----
[
  {
    "t1": {
      "airportname": "Grants Milan Muni",
      "airportname_dup": "high_altitude_Grants Milan Muni",
      "city": "Grants",
      "country": "United States",
      "faa": "GNT",
      "geo": {
        "alt": 6537,
        "lat": 35.167286,
        "lon": -107.901989
      },
      "icao": "KGNT",
      "id": 3439,
      "type": "airport",
      "tz": "America/Denver"
    }
  },
]

> ----

Reviewer: ```suggestion
.Results
[source,json]
----
[
  {
    "airportname_dup": "high_altitude_Grants Milan Muni"
  },
  {
    "airportname_dup": "high_altitude_Durango La Plata Co"
  },
  {
    "airportname_dup": "high_altitude_Black Rock"
  },
  ...
  {
    "airportname_dup": "high_altitude_Colorado Springs East"
  }
]
----
```

> == Example 10

Reviewer: Example 10 is incorrect and also needs results.

``` asciidoc
[source,n1ql]
----
UPDATE `travel-sample`.inventory.airport t1
SET airportname_dup = "high_altitude_2 " || airportname
WHERE (SELECT RAW geo.alt
       FROM t1.geo
       WHERE geo.alt > 6000)[0] = t1.geo.alt
RETURNING t1.airportname_dup;
----

.Results
[source,json]
----
[
  {
    "airportname_dup": "high_altitude_2 Grants Milan Muni"
  },
  {
    "airportname_dup": "high_altitude_2 Durango La Plata Co"
  },
  {
    "airportname_dup": "high_altitude_2 Black Rock"
  },
  ...
  {
    "airportname_dup": "high_altitude_2 Colorado Springs East"
  }
]
----
```

> 

Reviewer: ```suggestion

If you are warned that the query contains no WHERE clause or USE KEYS, choose btn:[Continue].

```

Reviewer: Missed the note about the warning that the Query Workbench gives you when using the UPDATE command.

> [source,json]

Reviewer: Query is N1QL

```suggestion
[source,n1ql]
```

----
UPDATE `travel-sample`.inventory.hotel t1
SET reviews_5star = (SELECT raw t2
                     FROM t1.reviews t2
                     WHERE t2.ratings.Overall = 5
                     ORDER BY t2.author)
LIMIT 1
RETURNING t1.reviews[*].author, t1.reviews1;
----

> ----

Reviewer: Result is JSON

```suggestion
----

.Results
[source,json]
----
```

[
  {
    "author": [
      "Ozella Sipes",
      "Barton Marks"
    ]
  }
]
----

> == Example 12

Author: @simon-dew this is where I am stuck at. Need your help

----
INSERT INTO `travel-sample`  t1 (KEY _k, VALUE _v)
SELECT "newkey_" || meta(t2).id as _k, t2.airportname as _v
FROM `travel-sample`.inventory.airport t2
WHERE t2.geo.alt > 6400
RETURNING _k;

> ----

Reviewer: This example does not contain a subquery, correlated or otherwise. Replace this entire example code block with a different example from the INSERT page, and include the results. Don't forget that the language of the source code block should be N1QL, not JSON!

``` asciidoc
[source,n1ql]
----
INSERT INTO `travel-sample`.inventory.hotel (KEY UUID()) 
    SELECT x.name, x.city, "landmark_hotels" AS type 
      FROM `travel-sample`.inventory.hotel x
      WHERE x.city WITHIN
        ( SELECT DISTINCT t.city 
            FROM `travel-sample`.inventory.landmark t)
      LIMIT 4
RETURNING *;
----

.Results
[source,json]
----
[
  {
    "hotel": {
      "city": "Aberdeenshire",
      "name": "Castle Hotel",
      "type": "landmark_hotels"
    }
  },
  {
    "hotel": {
      "city": "Aberdeenshire",
      "name": "Two Bears Cottage",
      "type": "landmark_hotels"
    }
  },
  {
    "hotel": {
      "city": "Agoura Hills",
      "name": "Malibu Creek Campground",
      "type": "landmark_hotels"
    }
  },
  {
    "hotel": {
      "city": "Altrincham",
      "name": "Cresta Court Hotel",
      "type": "landmark_hotels"
    }
  }
]
----
```

> [source,json]

Reviewer: This query won't do anything because there are no reviews with an overall rating of 0! 1 is the lowest overall rating. First, please change the prose description of this example above to say:

```
A correlated subquery with DELETE.
This example deletes all hotels which got the lowest overall rating from more than 4 reviewers.
```

Then change the language of the query code block to N1QL:

```suggestion
[source,n1ql]
```

----
DELETE FROM `travel-sample`.inventory.hotel t
WHERE (SELECT RAW count(*)
                          FROM t.reviews t2

>                           WHERE t2.ratings.Overall = 0 )[0] > 4;

Reviewer: Update the query. Let's also return the hotel name so the user can see what's happened.

```suggestion
DELETE FROM `travel-sample`.inventory.hotel t
WHERE (SELECT RAW count(*)
       FROM t.reviews t2
       WHERE t2.ratings.Overall = 1 )[0] > 4
RETURNING t.name;
```

> ----

Reviewer: Finally, show the expected results.

```suggestion
----

.Results
[source,json]
----
[
  {
    "name": "Beverly Laurel Motor Hotel"
  },
  {
    "name": "Tan yr Eglwys Cottages"
  }
]
----
```

In this example, the subquery filters for 5 rated reviews and sorts them by reviewer name.
The result of the subquery is assigned to a new field `reviews_5star` in the hotel document.

[source,n1ql]
----
UPDATE `travel-sample`.inventory.hotel t1
SET reviews_5star = (SELECT raw t2
                     FROM t1.reviews t2
                     WHERE t2.ratings.Overall = 5
                     ORDER BY t2.author)
LIMIT 1

> RETURNING t1.reviews[*].author, t1.reviews1;

Reviewer: `t1.reviews1` is meaningless. I must have missed this one.

```suggestion
RETURNING t1.reviews[*].author;
```

----

.Results
[source,json]
----
[
  {
    "author": [
      "Ozella Sipes",
      "Barton Marks"
    ]
  }
]
----

== Example 12

A non-correlated subquery with INSERT.

[source,n1ql]
----
INSERT INTO `travel-sample`.inventory.hotel (KEY UUID()) 
    SELECT x.name, x.city, "landmark_hotels" AS type 
      FROM `travel-sample`.inventory.hotel x
      WHERE x.city WITHIN
        ( SELECT DISTINCT t.city 
            FROM `travel-sample`.inventory.landmark t)
      LIMIT 4
RETURNING *;
----

.Results
[source,json]
----
[
  {
    "hotel": {
      "city": "Aberdeenshire",
      "name": "Castle Hotel",
      "type": "landmark_hotels"
    }
  },
  {
    "hotel": {
      "city": "Aberdeenshire",
      "name": "Two Bears Cottage",
      "type": "landmark_hotels"
    }
  },
  {
    "hotel": {
      "city": "Agoura Hills",
      "name": "Malibu Creek Campground",
      "type": "landmark_hotels"
    }
  },
  {
    "hotel": {
      "city": "Altrincham",
      "name": "Cresta Court Hotel",
      "type": "landmark_hotels"
    }
  }
]
----

== Example 13

> A correlated subquery with DELETE all hotel records which got zero overall rating by more than 4 reviewers.

Reviewer: There are no zero ratings. The minimum is 1.

```suggestion
A correlated subquery with DELETE all hotel records which got the minimum overall rating by more than 4 reviewers.
```
